graph.py

```diff
@@ -17,12 +41,110 @@ def __init__(self, builds):
         for build in builds:
             self.add(build)
 
+    def _get_layer_info(self, layer_id, pulp_repo_url):
+        try:
+            size = None
+            nvr = None
+            parent_id = None
+            layer_json = {}
+            if layer_id in self.known_pulp_layers.keys():
+                nvr = self.known_pulp_layers[layer_id]
+                size = self.found_nvr_sizes[nvr]
+            else:
+                pulp_url = pulp_repo_url + '/%s/json' % layer_id
+                r = requests.get(pulp_url, verify=False)
+                r.raise_for_status()
+                layer_json = r.json()
+        except Exception as e:
+            sys.stderr.write("  _get_layer_info(%s): %s" % (layer_id, repr(e)))
+            return (None, None, None)
+
+        try:
+            size = layer_json.get('Size', 0)
+            parent_id = layer_json.get('parent_id', None)
+            labels = layer_json['config'].get('Labels', {'Name': '', 'Version': '', 'Release': ''})
+            nvr = '%s-%s-%s' % (labels['Name'], labels['Version'], labels['Release'])
+        except Exception as e:
+            sys.stderr.write("  _get_layer_info(%s): %s" % (layer_id, repr(e)))
+
+        return (parent_id, size, nvr)
+
+    def _get_upload_size(self, build):
+        annotations = build['metadata']['annotations']
+        image_id = annotations['image-id']
+        if image_id in self.pulp_upload_size.keys():
+            return self.pulp_upload_size[image_id]
+
+        repos_json = json.loads(annotations['repositories'])
+        unique_repos = repos_json['unique']
+        if not unique_repos:
+            self.pulp_upload_size[image_id] = 0
+            return 0
+
+        repo = repos_json['primary'][0]
+        expected_nvr = "%s/%s-%s-%s" % (
+            repo.split('/')[-2],
+            repo.split('/')[-1].split(':')[0],
+            repo.split('/')[-1].split(':')[1].split('-')[0],
+            '-'.join(repo.split('/')[-1].split(':')[1].split('-')[1:])
+        )
+        if image_id in self.known_pulp_layers.keys():
+            size = self.found_nvr_sizes[expected_nvr]
+            self.pulp_upload_size[image_id] = size
+            sys.stderr.write("  found cached size for %s: %s\n" % (image_id, size))
+            return size
+
+        full_image_name = unique_repos[0]
+        image_name = '-'.join(strip_registry_from_image(full_image_name).split('/'))
+        pulp_repo_url = '%s/pulp/docker/v1/redhat-%s' % (BASE_PULP_URL, image_name.split(':')[0])
```

Comment:
That can happen if the image was removed from Pulp since being built (e.g. it was a beta image no longer in beta, or dropped). I think in that case there should just be a warning.